### PR TITLE
Fix format of self identifier line with AlternativeLongMemberDefinitions

### LINKED ISSUE #2598: AlternativeLongMemberDefinitions breaks spacing around `type Blah as this`
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAyogE4BuxAPAOQCGAfADoB2EEAFEy52NADYCuAZ3TF8ASwBeeNKMboANBy4RcRAaKGJZ%2BdNWFg0ASQAKxNRq07hVOguZKwAW0QCB1AOaIAwgAtqjRog8%2BmAy6OSEpMQG8LRgALSxALLOrh7WsQlgfIyi6IoQAJT5ENQCYOje6gC8xTyI6I58OgBGdSFgVWAADMVODs3E5ZUCAHQAQjzU3h3dkHLgUAL16DJuAqhgANoLEKBKkAAkWLgbkKFzOyyHJNT8yGjb9t3zT5AGjEhaEsgv9gAslwgAF1LgBfX67QFQI44e5Qc4Qq4HG53DaPewARkRnEgiWoAA8ADIyRCEzRuCoXJ4ADi6UJBnHBlz2XEOxzhkE08AA8gAzYkBbSYOpU1nI258OHopQAJmxVwAoh8%2BQKflCIJAeLzIPSwYiWTiYSc0JBmrBYDxRYaUZK0eqwABmeUa95LIjoABiMluADlEAB3Hgkq1cdBESW6xn6qFs2GnM0WkMa8Woh72v7OqD4bDUaCIMaIXmwIiIIzUIjUJzCIhJlhhiNKBksJmcA1I9nx82W%2BXXCVS%2B0AVkzkGzufzheLpNg-uI0FKiHeJFgc5WsEYtYg9cQkeb0f2Ro5Ce7MZTtrTTwAbMPR3mC0WSwBVbAqOdLRfL3SiNcbsC825LHcIBbFg22TDsTSPWte1TLZ7QAdmvHNbwnEtPEmFxPDXIRw2gdBix-P8eAAxs9WZE9wKgSCe1PfsaUQsc70nZJ%2BmIAj-23EiozI-cKNNLsoJou0ngATno5D7y8C1v0zQjiK4JsgL3MVeKok8bVozEujExAAEFeWrTCHAcagfy3QCwGAyEeLjCD%2BOo9ShMxLF7RHJDxwkwgHFEOAeGk%2B1ZI4%2BTSNbcibMouy1L7RylAxOUXJvXT9JERAvJ8vynjMzjd245Swr4xN7Ki89MSdeK3J0ohYGyeAABFAlELzq1M8NAs4BSLKUw1eIRSKYOlLgMQzFy8XxAxeQAFW8TR8G8Yt0AAdVEeBKUzOkssUnKurynr9wc4qYqHYaCTGybNAVIjEBmubFuW7wfwvNagq4kLrONeFZAEvbYKeDEryO0bGF5UR8W5FRdGLBV8WwEsXC-ddM1pczLLAUDoW6j7Cr6%2B0MQQ-6ACVEDgIh4Bulb00etrgpA0K3rODHerPb7MWpYcRoJon4B9PgWKIPkDGEBx1kzDEkc69s8ocPgeBWIMAg9YtjPQatxrjTHGf6zgMVElz2eLeBEilmWSXlohFaa4doF8CtcOIAB9f0lspUXNvF2mdrFL6NZYGUtP%2BiqK0wbkiGJIRSbuhGKZYdrkdR2M3fp3aiqZ2VnNeEb-eoQPg-UdAuZ5vmBaF7HnZe3Lacl6XRFlxATbN4gVZONWNNlOLXgzrOQ-QA3K%2Br2vdHNlzLfLXNq3tx27pL6nXo5d3rSTr2IBlUq04JAA1PsxhkeBVjDn9EfWjqXbA7aE49%2Bf7RlIaV-xD1slwuHN4%2BHfx5-P5I%2BBKmrLLmfT7nrGnhlIda%2BNVYDoAAOL1EhtDFIcNd4R0nl-LatNVKJ3-vYGUf006GyriSCYy4ADWYxrb4PqAIbkjB8CVkkvwBw8N-LsQQSjGmh4IqoPVhfXGrw-SBlwfUGcmgG6IDqkDHIq5GA6Q%2BMxAYqhmoNietlUuSCWEFQZs3LgMoWYuR0kGNwjBb6MHvmuMQujdB8BLONWA7xPiWDhrI1qUdP5MOnp2FRbC1GcBlNrNu0tiCME-GQQka43BSOIMImQuQ4ZFwyi1RhscDwuOPG46KXAHS%2BywZXVUhJKzNHgNQNCsA1CMDcNw6ubELqxOYQkz658ngOlTvYSAUDiANU0DoHgABpRAiBsBWNae8IhfhLZlLkpTZ6U9v5VKbskzgDpW4NImH4fBqoBAVSqh8P0Gh9bYOrlAmGagsJ2Iqc42yriz5oKUA6Ze8zyyMRLDVdQ0AiANW9MIeAD4chrjqrwYeYjhn2I-mMxBrtlGJLOew2pV8GlNKeU4WQtxtCVUEPWbAQoRQyQYQfGOlSTSzyRJ7e0DogENM6d0kaedpF8gWYwJZJIomaXfmAaOYtj7INYWC9xLAHSYIaQi7yXcEBqieAFJGCwgS-DOAID0agNjCtBEAA)

#### Code

```fsharp
type Server<'a>
    (
        clusterSize : int,
        persistentState : IPersistentState<'a>,
        messageChannel : int<ServerId> -> Message<'a> -> unit
    )
     as this=
    let mutable i = 0
    member this.Blah = 0
```
            

#### Result

```fsharp
type Server<'a>
    (
        clusterSize: int,
        persistentState: IPersistentState<'a>,
        messageChannel: int<ServerId> -> Message<'a> -> unit
    )
     as this=
    let mutable i = 0
    member this.Blah = 0

```
            
#### Problem description

A stylistic bug only. Enabling `AlternativeLongMemberDefinitions` caused the strange spacing you see on the `as this=` line; I would expect it to format in the way it does without that setting, i.e.:

```fsharp
type Server<'a>
    (
        clusterSize: int,
        persistentState: IPersistentState<'a>,
        messageChannel: int<ServerId> -> Message<'a> -> unit
    ) as this =
    let mutable i = 0
    member this.Blah = 0
```

#### Extra information

- [ ] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.

#### Options

Fantomas main branch at 2022-10-24T07:25:37Z - 529aa7d7807a3b996045c08083a2635f9588a37e

```fsharp
    { config with
                MaxLineLength = 80
                AlignFunctionSignatureToIndentation = true
                AlternativeLongMemberDefinitions = true }
```

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://fsprojects.github.io/fantomas/docs/end-users/IgnoreFiles.html)?</sub>

### PR DESCRIPTION
**Pull request by dawedawe**
fixes #2598

### CODE REVIEW

File: src/Fantomas.Core/CodePrinter.fs [hunk 1]
```diff
@@ -3110,7 +3110,7 @@ and genTypeDefn (TypeDef (ats, px, leadingKeyword, ao, tds, tcs, equalsRange, td
             (opt sepNone impCtor (fun mdf -> sepSpaceBeforeClassConstructor +> genMemberDefn mdf))
             (fun isMultiline ctx ->
                 if ctx.Config.AlternativeLongMemberDefinitions && isMultiline then
-                    genEqFixed SynTypeDefn_Equals equalsRange ctx
+                    sepSpace ctx |> genEqFixed SynTypeDefn_Equals equalsRange
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:3113]: I would be a bit more explicit here.
You can check the presence of the implicit ctor alias:

```fsharp
let hasAliasInImplicitConstructor =
    match impCtor with
    | Some(SynMemberDefn.ImplicitCtor(selfIdentifier = Some _)) -> true
    | _ -> false
```

and then just tighten the condition:
```fsharp
if ctx.Config.AlternativeLongMemberDefinitions && isMultiline && not hasAliasInImplicitConstructor then
    genEqFixed SynTypeDefn_Equals equalsRange ctx
else
    genEq SynTypeDefn_Equals equalsRange ctx
```

File: src/Fantomas.Core/CodePrinter.fs [hunk 2]
```diff
@@ -3923,7 +3923,12 @@ and genMemberDefn node =
         // In implicit constructor, attributes should come even before access qualifiers
         ifElse ats.IsEmpty sepNone (sepSpace +> genOnelinerAttributes ats)
         +> genCtor
-        +> optPre (!- " as ") sepNone so genIdent
+        +> fun ctx ->
```
Comment (nojaf) [inline src/Fantomas.Core/CodePrinter.fs:3926]: A simpler check would be to use `sepSpace` like `optPre (sepSpace +> !- "as ") sepNone so genIdent`. The same check is already encapsulated there.

**Review by nojaf — COMMENTED** [review]
Thanks for the quick fix.
I would approach things slightly differently.
Let me know what you think.

**Comment by dawedawe** [comment]
> Thanks for the quick fix. I would approach things slightly differently. Let me know what you think.

Thanks, I still have problems choosing the simplest approach with our infrastructure.

**Comment by nojaf** [comment]
Yeah, it can be quite hard to see the simplest solution at first glance.
I often see some of my older code and frown at why such a complex solution was used.

**Review by nojaf — APPROVED** [review]

**Pull request merged**